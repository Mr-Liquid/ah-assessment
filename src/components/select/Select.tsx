import { useState, useRef } from 'react';
import { useClickOutside } from '../../hooks/useClickOutside/useClickOutside';
import { SelectOption } from './SelectOption';
import { ArrowIcon } from '../icons';

type SelectProps = {
  value?: string;
  options?: { label: string; value: string }[];
  onChange?: (value: string) => void;
};

const Select = ({
  value = '',
  options,
  onChange = () => null,
}: SelectProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const label = options?.find(option => option.value === value)?.label ?? '';
  const [selectedOption, setSelectedOption] = useState<string | ''>(label);
  const innerRef = useRef<HTMLDivElement | null>(null);

  const toggleDropdown = () => setIsOpen(!isOpen);

  const handleOptionClick = (value: string) => {
    const label = options?.find(option => option.value === value)?.label ?? '';
    onChange(value);
    setSelectedOption(label);
    setIsOpen(false);
  };

  useClickOutside(innerRef, () => setIsOpen(false));

  const handleButtonClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    toggleDropdown();
  };

  return (
    <div className="flex items-center">
      <div className="relative inline-block text-left" ref={innerRef}>
        <div>
          <button
            data-testid="toggle-select-btn"
            id="dropdown-button"
            type="button"
            onClick={handleButtonClick}
            className="inline-flex justify-center items-center w-60 rounded-md border-2 border-gray-200 shadow-sm px-4 py-2 bg-white-700 text-sm text-gray-600 hover:bg-white-600 focus:outline-none"
            aria-haspopup="true"
            aria-expanded={isOpen}
          >
            <span
              id="dropdown-selected-option"
              className="w-full text-left overflow-hidden flex-1"
            >
              {selectedOption === '' ? 'Select Year' : selectedOption}
            </span>
            <ArrowIcon
              className="ml-2.5 -mr-1.5 h-5 w-5"
              style={{ transform: isOpen ? 'rotate(180deg)' : 'rotate(0deg)' }}
            />
          </button>
        </div>
        {isOpen && (
          <div
            id="dropdown-menu"
            data-testid="dropdown-menu"
            className="overflow-scroll h-[200px] bg-white origin-top-right absolute w-full left-0 mt-2 rounded-md shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none"
            role="listbox"
            aria-orientation="vertical"
            aria-labelledby="dropdown-button"
            tabIndex={-1}
          >
            {options?.map((option, index) => {
              return (
                <SelectOption
                  key={`option.label-${index}`}
                  option={option}
                  handleOptionClick={handleOptionClick}
                />
              );
            })}
          </div>
        )}
      </div>
    </div>
  );
};

export { Select };
