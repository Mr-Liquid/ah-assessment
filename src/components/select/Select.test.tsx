import { render, screen, fireEvent } from '@testing-library/react';
import { Select } from './Select';
import { describe, it, expect } from 'vitest';
import { act } from 'react';

describe('Select Component', () => {
  it('matches snapshot with props', () => {
    const { container } = render(
      <Select options={[{ label: 'Option 1', value: '1' }]} value="1" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('toggles dropdown on button click', () => {
    render(
      <Select
        options={[
          { label: '2021', value: '2021' },
          { label: '2022', value: '2022' },
        ]}
      />
    );

    expect(screen.queryByTestId(/dropdown-menu/i)).not.toBeInTheDocument();
    const toggleButton = screen.getByTestId(/toggle-select-btn/i);
    fireEvent.click(toggleButton);

    act(() => {
      expect(screen.getByTestId(/dropdown-menu/i)).toBeInTheDocument();
    });

    fireEvent.click(toggleButton);

    act(() => {
      expect(screen.queryByTestId(/dropdown-menu/i)).not.toBeInTheDocument();
    });
  });

  it('closes dropdown on outside click', () => {
    render(
      <div>
        <Select
          options={[
            { label: '2021', value: '2021' },
            { label: '2022', value: '2022' },
          ]}
        />
        <button data-testid="outside-button">Outside</button>
      </div>
    );

    const toggleButton = screen.getByTestId(/toggle-select-btn/i);
    const outsideButton = screen.getByTestId(/outside-button/i);

    fireEvent.click(toggleButton);
    act(() => {
      expect(screen.getByTestId(/dropdown-menu/i)).toBeInTheDocument();
    });

    fireEvent.click(outsideButton);
    act(() => {
      expect(screen.queryByTestId(/dropdown-menu/i)).not.toBeInTheDocument();
    });
  });
});
