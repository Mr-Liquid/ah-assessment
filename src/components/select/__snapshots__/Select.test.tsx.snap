// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select Component > matches snapshot with props 1`] = `
<div
  class="flex items-center"
>
  <div
    class="relative inline-block text-left"
  >
    <div>
      <button
        aria-expanded="false"
        aria-haspopup="true"
        class="inline-flex justify-center items-center w-60 rounded-md border-2 border-gray-200 shadow-sm px-4 py-2 bg-white-700 text-sm text-gray-600 hover:bg-white-600 focus:outline-none"
        data-testid="toggle-select-btn"
        id="dropdown-button"
        type="button"
      >
        <span
          class="w-full text-left overflow-hidden flex-1"
          id="dropdown-selected-option"
        >
          Option 1
        </span>
        <svg
          class="cursor-pointer  ml-2.5 -mr-1.5 h-5 w-5"
          fill="currentColor"
          height="21"
          stroke="#8899a4"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1"
          style="transform: rotate(0deg);"
          width="21"
        >
          <path
            clip-rule="evenodd"
            d="M6.293 7.293a1 1 0 011.414 0L10 9.586l2.293-2.293a1 1 0 111.414 1.414l-3 3a1 1 0 01-1.414 0l-3-3a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
